### DISCUSSION TITLE
up and down arrows keys not working in shell

### DISCUSSION THREAD

**Discussion by ws1088** [python-poetry/poetry#3760] [Q&A]
when i start a shell (Windows 10) with `poetry shell`, the up and down arrow keys are not working. they are supposed to go to previous commands i executed. is there something i forgot to do?

**Comment by LiYueqian-James** [comment]
Had the same issue, did u find any solutions?

**Reply by iwatanab** [reply]
Same issue here. Has anyone solved this yet?

**Reply by bmarroquin** [reply]
Are you using cmd.exe or powershell.exe? Are you using pyenv for windows?

I ask because this works for me on powershell. Could be a cmd.exe problem.

**Reply by bridgeL** [reply]
> Are you using cmd.exe or powershell.exe? Are you using pyenv for windows?
> 
> I ask because this works for me on powershell. Could be a cmd.exe problem.

Well, for me, things are in another case. I'm using **powershell** and run `poetry shell` to switch into the **poetry shell**. In this stage, I can use up arrow to get previous input. However, when I run `python`, then in the python I can't use my up arrow while if I run `python` out of  **poetry shell**, it can work

**Comment by bridgeL** [comment]
I find a solution and it does work.

https://www.reddit.com/r/django/comments/tzj0ig/python_interpreter_history_doesnt_work_updown/

> Well, figured it out. Sounds like the proposed fix didn't make it into 1.2.0b1, but can be run manually. So future people who find this issue, the fix can be found at: https://discuss.python.org/t/interactive-command-history-in-session-started-with-subprocess-on-windows/3701/5
> When you launch, put that in a file and call set_console_history_info() and you'll be all good.

**Comment by i-d-lytvynenko** [comment]
For anyone dealing with the same issue:

If [changing the cmd options](https://github.com/python-poetry/poetry/pull/5053#issuecomment-1015640706) doesn't fix the problem (for example, in cmd inside wezterm or neovim, as your cmd options are ignored in such cases) and you don't want to edit the poetry files (as suggested [here](https://github.com/orgs/python-poetry/discussions/3760#discussioncomment-3991244)), one may compile the following program:
```c
#include <windows.h>

int main() {
    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);

    CONSOLE_HISTORY_INFO historyInfo;
    historyInfo.cbSize = sizeof(CONSOLE_HISTORY_INFO);
    historyInfo.HistoryBufferSize = 1000;
    historyInfo.NumberOfHistoryBuffers = 10;

    SetConsoleHistoryInfo(&historyInfo);

    return 0;
}
```
and run it with cmd like so: `cmd /k D:/path/to/change_max_buffers.exe`

For example, the solution for neovim looks like:
```lua
vim.cmd([[autocmd TermOpen * call chansend(b:terminal_job_id, "D:/path/to/change_max_buffers.exe\r")]])
```

**Comment by thekashifmalik** [comment]
I have this issue on Linux. It's not just up / down but all commands like `ctrl + p`, `ctrl + n`, etc. They spit out escape codes instead.

**Comment by ragumagu** [comment]
If you just need a python shell (running with env created with poetry), and If you don't want to run scripts to enable history, you can directly run python.exe without using poetry shell:

```
Invoke-Expression ($(poetry env info --path) + "\Scripts\python.exe")
```

This will get the arrow keys working, just not spawning a new shell.
